python/openchange/migration/directory.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# OpenChangeDB DB schema and its migrations
+# Copyright (C) Enrique J. Hernández Blasco <ejhernandez@zentyal.com> 2015
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Migration for OpenChange directory schema and data
+"""
+from openchange.migration import migration, Migration
+import ldb
+
+from samba.samdb import SamDB
+from samba.auth import system_session
+from samba.provision import setup_modify_ldif
+from samba.net import Net
+from samba.dcerpc import nbt
+
+# this helper functions are extracted from openchange.provision to avoid circular import
+# it would be better a refactoring of provision.ppy so we can avoid repeating code here
+
+
+def get_ldb_url(lp, creds, names):
+    if names.serverrole == "member server":
+        net = Net(creds, lp)
+        dc = net.finddc(domain=names.dnsdomain, flags=nbt.NBT_SERVER_LDAP)
+        url = "ldap://" + dc.pdc_dns_name
+    else:
+        url = lp.samdb_url()
+
+    return url
+
+
+def get_local_samdb(names, lp, creds):
+    samdb_url = get_ldb_url(lp, creds, names)
+    session_info = system_session()
+    local_db = SamDB(samdb_url, session_info=session_info,
+                     credentials=creds, lp=lp)
+    return local_db
+
+
+def force_schemas_update(sam_db, setup_path):
+    sam_db.transaction_start()
+    try:
+        # update schema now
+        setup_modify_ldif(sam_db, setup_path('AD/update_now.ldif'), None)
+    except:
+        sam_db.transaction_cancel()
+        raise
+    sam_db.transaction_commit()
+
+
+# end of helper functions
+
+@migration('directory', 1)
+class AddRecipientAttributes(Migration):
+    description = 'Add missing recipient attributes'
+
+    @classmethod
+    def apply(cls, cur, extra):
+        lp = extra['lp']
+        creds = extra['creds']
```

Comment:
As I stated above, this is undocumented. And it must be, using `**kwargs` approach you could use it here as normal parameters like `(..., cur, lp, creds, names, setup_path)`